### ISSUE TITLE
Signal handling error in daemon mode

### ISSUE THREAD

**Issue by eunsoopark** [droe/sslsplit#110]
In daemon mode, sslsplit can't process any traffic when receive SIGPIPE.

In debug mode, it works well.

following is sslsplit status when receive SIGPIPE

sslsplit process STAT changed from Ssl to Rsl and stop with futex lock.

bash-3.1# ps x|grep sslsplit
 7384 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /tmp/CA.crt -k /tmp/CA.prv ssl 0.0.0.0 40022 netfilter
 7577 pts/0    S+     0:00 grep sslsplitbash-3.1# killall -PIPE sslsplit
bash-3.1# ps x
bash-3.1# ps x|grep sslsplit
 7384 ?        Rsl    0:01 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /tmp/CA.crt -k /tmp/CA.prv ssl 0.0.0.0 40022 netfilter
 7641 pts/0    S+     0:00 grep sslsplit
bash-3.1# strace -f -p 7384
Process 7384 attached with 66 threads
[pid  7449] epoll_wait(402,  <unfinished ...>
[pid  7448] epoll_wait(396,  <unfinished ...>
[pid  7447] epoll_wait(390,  <unfinished ...>
[pid  7446] epoll_wait(384,  <unfinished ...>
[pid  7445] epoll_wait(378,  <unfinished ...>
[pid  7444] epoll_wait(372,  <unfinished ...>
[pid  7443] epoll_wait(366,  <unfinished ...>
[pid  7442] epoll_wait(360,  <unfinished ...>
[pid  7441] epoll_wait(354,  <unfinished ...>
[pid  7440] epoll_wait(348,  <unfinished ...>
[pid  7439] epoll_wait(342,  <unfinished ...>
[pid  7438] epoll_wait(336,  <unfinished ...>
[pid  7437] epoll_wait(330,  <unfinished ...>
[pid  7436] epoll_wait(324,  <unfinished ...>
[pid  7435] epoll_wait(318,  <unfinished ...>
[pid  7434] epoll_wait(312,  <unfinished ...>
[pid  7433] epoll_wait(306,  <unfinished ...>
[pid  7432] epoll_wait(300,  <unfinished ...>
[pid  7431] epoll_wait(294,  <unfinished ...>
[pid  7430] epoll_wait(288,  <unfinished ...>
[pid  7429] epoll_wait(282,  <unfinished ...>
[pid  7428] epoll_wait(276,  <unfinished ...>
[pid  7427] epoll_wait(270,  <unfinished ...>
[pid  7426] epoll_wait(264,  <unfinished ...>
[pid  7425] epoll_wait(258,  <unfinished ...>
[pid  7424] epoll_wait(252,  <unfinished ...>
[pid  7423] epoll_wait(246,  <unfinished ...>
[pid  7422] epoll_wait(240,  <unfinished ...>
[pid  7421] epoll_wait(234,  <unfinished ...>
[pid  7420] epoll_wait(228,  <unfinished ...>
[pid  7419] epoll_wait(222,  <unfinished ...>
[pid  7418] epoll_wait(216,  <unfinished ...>
[pid  7417] epoll_wait(210,  <unfinished ...>
[pid  7416] epoll_wait(204,  <unfinished ...>
[pid  7415] epoll_wait(198,  <unfinished ...>
[pid  7414] epoll_wait(192,  <unfinished ...>
[pid  7413] epoll_wait(186,  <unfinished ...>
[pid  7412] epoll_wait(180,  <unfinished ...>
[pid  7411] epoll_wait(174,  <unfinished ...>
[pid  7410] epoll_wait(168,  <unfinished ...>
[pid  7409] epoll_wait(162,  <unfinished ...>
[pid  7408] epoll_wait(156,  <unfinished ...>
[pid  7407] epoll_wait(150,  <unfinished ...>
[pid  7406] epoll_wait(144,  <unfinished ...>
[pid  7405] epoll_wait(138,  <unfinished ...>
[pid  7404] epoll_wait(132,  <unfinished ...>
[pid  7403] epoll_wait(126,  <unfinished ...>
[pid  7402] epoll_wait(120,  <unfinished ...>
[pid  7401] epoll_wait(114,  <unfinished ...>
[pid  7400] epoll_wait(108,  <unfinished ...>
[pid  7399] epoll_wait(102,  <unfinished ...>
[pid  7398] epoll_wait(96,  <unfinished ...>
[pid  7397] epoll_wait(90,  <unfinished ...>
[pid  7396] epoll_wait(84,  <unfinished ...>
[pid  7395] epoll_wait(78,  <unfinished ...>
[pid  7394] epoll_wait(72,  <unfinished ...>
[pid  7393] epoll_wait(66,  <unfinished ...>
[pid  7392] epoll_wait(60,  <unfinished ...>
[pid  7391] epoll_wait(54,  <unfinished ...>
[pid  7390] epoll_wait(48,  <unfinished ...>
[pid  7389] epoll_wait(42,  <unfinished ...>
[pid  7388] epoll_wait(36,  <unfinished ...>
[pid  7387] epoll_wait(30,  <unfinished ...>
[pid  7386] epoll_wait(24,  <unfinished ...>
[pid  7385] futex(0x17b41bc, FUTEX_WAIT_PRIVATE, 1, NULL

bash-3.1# ps -eLf x|grep sslsplit
root      7384     1  7384 95   66 17:18 ?        Rsl    5:32 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7385  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7386  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7387  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7388  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7389  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7390  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7391  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7392  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7393  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7394  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7395  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7396  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7397  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7398  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7399  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7400  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7401  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7402  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7403  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7404  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7405  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7406  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7407  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7408  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7409  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7410  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7411  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7412  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7413  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7414  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7415  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7416  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7417  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7418  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7419  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7420  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7421  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7422  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7423  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7424  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7425  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7426  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7427  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7428  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7429  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7430  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7431  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7432  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7433  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7434  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7435  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7436  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7437  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7438  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7439  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7440  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7441  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7442  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7443  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7444  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7445  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7446  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7447  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7448  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root      7384     1  7449  0   66 17:18 ?        Ssl    0:00 /usr/local/sbin/sslsplit -P -d -p /var/run/sslsplit.pid -u root -c /
root     11767 22841 11767  0    1 17:24 pts/0    S+     0:00 grep sslsplit

**Comment by droe** [comment]
To know what exact codebase you are using, what's the output of `sslsplit -V` on your system?

**Comment by eunsoopark** [comment]
bash-3.1# sslsplit -V
SSLsplit 0.4.11 (built 2015-11-09)
Copyright (c) 2009-2014, Daniel Roethlisberger <daniel@roe.ch>
http://www.roe.ch/SSLsplit
Build info: V:FILE
Features: -DHAVE_NETFILTER
NAT engines: netfilter* ahnlab tproxy ahnlab_tproxy
netfilter: IP_TRANSPARENT SOL_IPV6 !IPV6_ORIGINAL_DST
Local process info support: no
compiled against OpenSSL 1.0.1p 9 Jul 2015 (1000110f)
rtlinked against OpenSSL 1.0.1p 9 Jul 2015 (1000110f)
TLS Server Name Indication (SNI) supported
OpenSSL is thread-safe with THREADID
Using SSL_MODE_RELEASE_BUFFERS
SSL/TLS protocol availability: ssl3 tls10 tls11 tls12
SSL/TLS algorithm availability: RSA DSA ECDSA DH ECDH EC
OpenSSL option availability: SSL_OP_NO_COMPRESSION SSL_OP_NO_TICKET SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION SSL_OP_TLS_ROLLBACK_BUG
compiled against libevent 2.0.16-stable
rtlinked against libevent 2.0.16-stable
32 CPU cores detected

**Comment by droe** [comment]
I think the problem is that the signal handlers do not survive the forks part of detaching from TTY. The `develop` branch currently installs signal handlers after detaching from TTY, so this should not be an issue there. If you can test and confirm that, would be great.

Related to #104 in that the quickfix to #104 would probably reintroduce this bug in the `develop` branch.

**Comment by eunsoopark** [comment]
I have same result. 

Following is my change.

```diff
--- main.c  2015-11-11 09:54:50.760183018 +0900
+++ main.dev.c  2015-11-11 09:54:01.891263595 +0900
@@ -720,12 +720,14 @@
        exit(EXIT_FAILURE);
    }

+#if 0
    /* Bind listeners before dropping privileges */
    proxy_ctx_t *proxy = proxy_new(opts);
    if (!proxy) {
        fprintf(stderr, "%s: failed to initialize proxy.\n", argv0);
        exit(EXIT_FAILURE);
    }
+#endif
    /* Load certs before dropping privs but after cachemgr_preinit() */
    if (opts->tgcrtdir) {
        sys_dir_eachfile(opts->tgcrtdir, main_loadtgcrt, opts);
@@ -778,6 +780,13 @@

    rv = EXIT_SUCCESS;

+   /* Bind listeners before dropping privileges */
+   proxy_ctx_t *proxy = proxy_new(opts);
+   if (!proxy) {
+       fprintf(stderr, "%s: failed to initialize proxy.\n", argv0);
+       exit(EXIT_FAILURE);
+   }
+
    proxy_run(proxy);
    proxy_free(proxy);
    nat_fini();
```

**Comment by eunsoopark** [comment]
I solved this issue with libevent version update to 2.0.22.

I think following patch solved this issue.

https://github.com/libevent/libevent/commit/6e41cdc16b6e01840c0d54b5d7b09af3f1ed3cf4#diff-a49dae4bf6d105e1692f627817dc51af

This bug was patched libevent 2.0.17
